ste/userDelegationAuthenticationManager.go

```diff
@@ -0,0 +1,109 @@
+package ste
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/Azure/azure-storage-blob-go/azblob"
+)
+
+// userDelegationAuthenticationManager manages the automatic creation of user delegation SAS tokens for each container.
+// UDAM is only triggered when
+// 1) The user performs a service-to-service transfer
+// and
+// 2) The user did not supply a source SAS token on a blob transfer.
+//
+// When UDAM spins up, a credential is created.
+// UDAM is not necessary in the frontend because if UDAM would succeed, the OAuth token would have to be capable of enumerating the container to start with.
+type userDelegationAuthenticationManager struct {
+	// Because User Delegation SAS tokens cannot last longer than a user delegation credential, we must hold onto the key info.
+	startTime  time.Time
+	expiryTime time.Time
+	credential azblob.UserDelegationCredential
+	// This is a mostly-read scenario. Therefore, we treat the SAS map as an atomic value, and apply a write mutex to the map.
+	// We then, whenever we're attempting to get a SAS token for a container, check if it's present first.
+	// If it's not present, we start attempting to create the token. In order to do so, we lock the write mutex.
+	// Since this may take time, and by the time we obtain the lock, the SAS we're trying to get MAY have already been created,
+	// Check the map for the key before we overwrite.
+	sasMap           atomic.Value // points to a map[string]string. map[containerName]sasToken
+	sasMapWriteMutex sync.Mutex
+}
+
+// newUserDelegationAuthenticationManager uses an azblob service URL to obtain a user delegation credential, and returns a new userDelegationAuthenticationManager
+// serviceURL should have adequate permissions to generate a set of user delegation credentials.
+func newUserDelegationAuthenticationManager(serviceURL azblob.ServiceURL) (userDelegationAuthenticationManager, error) {
+	authManager := userDelegationAuthenticationManager{}
+	// Create a user delegation
+	authManager.startTime = time.Now()
+	authManager.expiryTime = authManager.startTime.Add(time.Hour * 24 * 7)
+
+	keyInfo := azblob.NewKeyInfo(authManager.startTime, authManager.expiryTime)
+
+	var err error
+	authManager.credential, err = serviceURL.GetUserDelegationCredential(steCtx, keyInfo, nil, nil)
```

Comment:
TODO: Consider dropping errors in their current use here, since errors out of UDAM are ignored anyway, and this is a holdover from the original design

Reply:
Also TODO: Consider refreshing the UDK, in case customers have long-running scenarios

Reply:
The latter is actually a humongous pain to do--  STE may have already taken expired SAS tokens if we don't refresh

A workaround is to just keep resuming the job whenever the token expires.

Reply:
WRT to this-- Perhaps passing around pointers to atomic.Values (and ensuring validity before we use a SAS token) would be the better solution in terms of handling SAS tokens in the STE if we need to refresh this.

Reply:
Although, thinking about it, this isn't a huge concern-- If somebody somehow runs into this, the suggestion is literally just "create a SAS token that lasts longer"

Reply:
Deciding NOT to handle things this way. As-is, despite the lack of use of errors, there _may_ be a use-case in the future.